### PR TITLE
added complex frame size support in framing #22129 (backport)

### PR DESCRIPTION
**Pull request by patriknw**
(cherry picked from commit b936ac641c9aaa993d054d414eca010cbef1f11f)

Refs #22129

### CODE REVIEW

**Review by ktoso — APPROVED** [review]
LGTM

File: akka-stream/src/main/scala/akka/stream/javadsl/Framing.scala [hunk 1]
```diff
@@ -94,6 +94,33 @@ object Framing {
...
+   *                         Then computeFrameSize can be used to compute the frame size: `(offset bytes, computed size) => (actual frame size)`.
+   *                         ''Actual frame size'' must be equal or bigger than sum of `fieldOffset` and `fieldLength`, the stage fails otherwise.
+   *
+   */
+  def lengthField(
+    fieldLength:        Int,
+    fieldOffset:        Int,
+    maximumFrameLength: Int,
+    byteOrder:          ByteOrder,
+    computeFrameSize:   akka.japi.function.Function2[Array[Byte], Int, Int]): Flow[ByteString, ByteString, NotUsed] =
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/javadsl/Framing.scala]: Whoop! Warning, should this not be Integers instead?

**Review by ktoso — CHANGES_REQUESTED** [review]
Whoop, signature in java method is bad here. This becomes `Function2<byte[], Object, Object>` in java (see comment in line)

File: akka-stream/src/main/scala/akka/stream/javadsl/Framing.scala (hunk 1 above)
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/javadsl/Framing.scala]: great catch

**Comment by patriknw** [comment]
this still needs another round of backporting when the original PR has been updated

**Comment by patriknw** [comment]
I have updated this with the latest from the other PR, so can be merged when validated.

**Comment by patriknw** [comment]
And for the record, this is perhaps not the kind of feature we would normally backport but it was a customer request.

**Pull request merged**